### DISCUSSION TITLE
`watch` is giving first item's value in the first render when there is 1 item in an array

### DISCUSSION THREAD

**Discussion by ykaragol** [react-hook-form/react-hook-form#3773] [Q&A]
**Describe the bug**

While using `useFieldArray` with an array of items, `watch` returns wrong value at the first run.

**To Reproduce**

1. Use `useFieldArray` for an array of items and have an initial list for `defaultValues`.
2. Remove all rows properly by clicking `delete` buttons for each row
3. Create new row after removing all rows.
4. See that `watch` is giving the value of the initial list at the first run.

**Codesandbox link (Required)**

[A sandbox link forked from docs](https://codesandbox.io/s/react-hook-form-usefieldarray-forked-q5nj1?file=/src/index.js)

0. See the first column calculating some data according second column
1. Delete all items by clicking delete
2. Append new one
3. See that there is a flicker on the first column that `Bill` is shown initially.

**Expected behavior**
It should not get that wrong value at initial render. (Because we want to have a calculation over that value and set as `defaultValue` to another input. But input gets wrong value.)

**Screenshots**
If sandbox not works, I can add.

**Desktop (please complete the following information):**

- OS: Mac OS
- Chrome
- Version: 6.11.0 (reproduced also in 6.13.1 too)

**Additional context**

**Comment by bluebill1049** [comment]
`watch` should not be used inside `render`, it's supposed to be invoked before `render`.

**Comment by ykaragol** [comment]
In the docs of `getValues`, it is saying:
> Important: You shouldn't use this method inside render.

But I didn't see anything about `watch`. 

How can I get the initial value of a field? Should I pass it from outside? But I don't want to have prop-drilling, can I get initial value from FormContext?

**Comment by bluebill1049** [comment]
there is an example under https://react-hook-form.com/api#watch
<img width="944" alt="Screen Shot 2020-12-28 at 9 23 52 am" src="https://user-images.githubusercontent.com/10513364/103180892-68a51e00-48ee-11eb-8ca5-5a66f7e09b2b.png">

```
import * as React from "react";
import { useForm, useFieldArray, ArrayField } from "react-hook-form";

function App() {
  const { register, control, handleSubmit, watch } = useForm();
  const { fields, remove, append } = useFieldArray({
    name: "test",
    control
  });
  const onSubmit = (data: FormValues) => console.log(data);
  
  console.log(watch("test")); 

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {fields.map((field, index) => {
        return (
          <input
            key={field.id}
            name={`test[${index}].firstName`}
            defaultValue={field.firstName}
            ref={register()}
          />
        );
      })}
      <button
        type="button"
        onClick={() =>
          append({
            firstName: "bill" + renderCount,
            lastName: "luo" + renderCount
          })
        }
      >
        Append
      </button>
    </form>
  );
}
```

**Comment by ykaragol** [comment]
Thank you @bluebill1049.

What I realized is that `console.log(watch("test"));` is working fine. But  `console.log(watch("test[0].firstName"));` is not.

In [the example you shared](https://codesandbox.io/s/watch-with-usefieldarray-542bu?file=/src/App.tsx), if I remove all items and append a new one, everything is working fine in console:
<img width="677" alt="Ekran Resmi 2020-12-28 10 22 37" src="https://user-images.githubusercontent.com/10911877/103197454-19afb680-48f7-11eb-87f5-9ec5de7724ab.png">


But when I updated `console.log(watch("test"));` as `console.log(watch("test[0]"));`; then console is showing this:
<img width="681" alt="Ekran Resmi 2020-12-28 10 23 11" src="https://user-images.githubusercontent.com/10911877/103197634-81660180-48f7-11eb-9334-07faf4e837f2.png">

Same issue happens in `console.log(watch("test[0].firstName"));`.

----

_Note: not worth mentioning but anyways: to make screenshots more readable, I've updated `console.log`s as `console.log(JSON.stringify(watch("test")));`._

**Comment by ykaragol** [comment]
Same issue happens when I `watch` it without any parameter:

```
  console.log(
    watch()
  );
```

Result:
<img width="750" alt="Ekran Resmi 2020-12-28 10 58 26" src="https://user-images.githubusercontent.com/10911877/103199312-a492b000-48fb-11eb-9936-7114a05a45d3.png">

**Reply by bluebill1049** [reply]
for Field Array, please watch the individual field array.

**Reply by ykaragol** [reply]
Is this by design? or isn't this an issue?

**Reply by ykaragol** [reply]
@bluebill1049 is this an issue to be fixed?

**Reply by bluebill1049** [reply]
There are extra logic we have baked for field array with `watch`, and yes it's by design.
